File: coding/src/com/algos/core/li04_twopointers/revised/L42_TrappingRainWater.java
package  com.algos.core.li04_twopointers.revised;

public class L42_TrappingRainWater {

	public static void main(String[] args) {
		L42_TrappingRainWater obj = new L42_TrappingRainWater();
		
		int[] height = new int[] {0,1,0,2,1,0,1,3,2,1,2,1};
		System.out.println(obj.trap(height));
	}
	
	// Problem: https://leetcode.com/problems/trapping-rain-water/
	// Idea: Track the maximum from left and right
	// If maxl is minimum, means the water that gets logged will be at max will be height of maxl. Use maxl to find the current position water logging maxl-height[i]
	// Similarly, update the water logging from the right if maxr is smaller then the maxl.
	public int trap(int[] height) {
        int sum = 0, n = height.length, maxl = height[0], maxr = height[n-1];
        int l=1, r=n-2;
        while(l<=r){
            if(maxl <= maxr){
                if(maxl-height[l] > 0){
                    sum += maxl-height[l];
                }
                maxl = Math.max(maxl, height[l]);
                l++;
            }else{
                if(maxr-height[r] > 0){
                    sum += maxr-height[r];
                }
                maxr = Math.max(maxr, height[r]);
                r--;
            }
        }
        
        return sum;
    }

}

File: dsa/patterns/03_twopointers/python/l42_trapping_rain_water.py
# L42 - Trapping Rain Water
#
# Problem: https://leetcode.com/problems/trapping-rain-water/
#
# Idea:
# - Use two pointers from both ends
# - Track max height from left (left_max) and right (right_max)
# - Water trapped depends on the smaller of the two maxes
#
# This is different from container with most water problem. Because in this
# we need to focus on addition of all waters trapped in all containers
# based on their respective heights in positions.
# Container problem focuses on finding max water container.
# Time: O(n)
# Space: O(1)


def trap(height):
    if len(height) < 3:
        return 0

    left, right = 0, len(height) - 1
    left_max = right_max = 0
    total_water = 0

    while left < right:
        if height[left] < height[right]:
            left_max = max(left_max, height[left])
            # If heigh is >, it will be update above already. So, no -ve addition.
            total_water += left_max - height[left]
            left += 1
        else:
            right_max = max(right_max, height[right])
            total_water += right_max - height[right]
            right -= 1

    return total_water


# -------------------- TEST --------------------
if __name__ == "__main__":
    height = [0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]
    print(trap(height))  # Output: 6
